### PR TITLE
Spider: find applet attrs, img dynsrc/lowsrc, and input src

### PR DESCRIPTION
**Pull request by Leyart**
Spider: find applet src, codebase
Spider: find img dynsrc
Spider: find img lowsrc
Spider: find input src

Part of #7149.

Signed-off-by: Luca Galli <leyart@gmail.com>

### CODE REVIEW

File: zap/src/test/resources/org/zaproxy/zap/spider/parser/html/ImgElementsSpiderHtmlParser.html [hunk 1]
```diff
@@ -22,6 +22,8 @@
 <img src="mailto:img@example.com">
 <img src="javascript:hello();">
 <img src="scheme://img.example.com/invalid">
+<img lowsrc="/img/lowresolution">
+<img dynsrc="https://video.example.com/native/video">
```
Comment (kingthorin) [inline zap/src/test/resources/org/zaproxy/zap/spider/parser/html/ImgElementsSpiderHtmlParser.html:26]: You seem to have added these to the set of URLs which are expected to be ignored.

File: zap/src/test/resources/org/zaproxy/zap/spider/parser/html/InputElementsSpiderHtmlParser.html [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+<head>
+<base href="http://example.com/sample/">
+<meta charset="UTF-8">
+<title>Input Elements - Spider HTML Parser</title>
+</head>
+<body>
+<!--  Ignored: -->
+<input>
+
+<!--  src -->
```
Comment (kingthorin) [inline zap/src/test/resources/org/zaproxy/zap/spider/parser/html/InputElementsSpiderHtmlParser.html:12]: Both comments seem to be preceded by double space.

File: zap/src/test/resources/org/zaproxy/zap/spider/parser/html/InputElementsSpiderHtmlParser.html [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+<input type="image" src="/relative/src" name="test" value="test">
```
Comment (kingthorin) [inline zap/src/test/resources/org/zaproxy/zap/spider/parser/html/InputElementsSpiderHtmlParser.html:13]: With a leading slash wouldn't this be an absolute path?

(Ex: Server root would be the leading slash.)

File: zap/src/test/resources/org/zaproxy/zap/spider/parser/html/AppletElementsSpiderHtmlParser.html [hunk 4]
```diff
@@ -0,0 +1,39 @@
+<!DOCTYPE html>
+<html>
+<head>
+<base href="http://example.com/sample/">
+<meta charset="UTF-8">
+<title>Input Elements - Spider HTML Parser</title>
```
Comment (kingthorin) [inline zap/src/test/resources/org/zaproxy/zap/spider/parser/html/AppletElementsSpiderHtmlParser.html:6]: Applet Elements?

File: zap/src/test/resources/org/zaproxy/zap/spider/parser/html/AppletElementsSpiderHtmlParser.html [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+<applet code="upgradedGame.class" align="left" src="https://example.com/game/applet/file.class" height="250" width="350">
+    <param name="difficulty" value="easy">
+    <b>Sorry, you need Java to play this upgraded game.</b>
+</applet>
+
+<applet code="codebase.class" align="left" codebase="https://example.com/absolute/applet/" height="250" width="350">
+</applet>
+
+<applet code="codebase.class" align="left" codebase="/relative/applet/" height="250" width="350">
```
Comment (kingthorin) [inline zap/src/test/resources/org/zaproxy/zap/spider/parser/html/AppletElementsSpiderHtmlParser.html:29]: With the leading slash this is absolute right? (Ex: from server root this is the full actual path.)

File: zap/src/test/resources/org/zaproxy/zap/spider/parser/html/ImgElementsSpiderHtmlParser.html (hunk 1 above)
Comment (Leyart) [reply zap/src/test/resources/org/zaproxy/zap/spider/parser/html/ImgElementsSpiderHtmlParser.html:26]: Thanks, overlooked :)

File: zap/src/test/resources/org/zaproxy/zap/spider/parser/html/InputElementsSpiderHtmlParser.html (hunk 2 above)
Comment (Leyart) [reply zap/src/test/resources/org/zaproxy/zap/spider/parser/html/InputElementsSpiderHtmlParser.html:12]: Thought it was a convention, also other comments are with double spaces. But can be fixed for clarity

**Comment by kingthorin** [comment]
Build failed due to code needing formatting. Please run `./gradlew spotlessApply`, commit (you can amend), and push (if you amend you'll need to `--force`).

```bash
./gradlew spotlessApply
git commit -a --amend --no-edit --signoff
git push --force
```

File: zap/src/test/resources/org/zaproxy/zap/spider/parser/html/ImgElementsSpiderHtmlParser.html (hunk 1 above)
Comment (kingthorin) [reply zap/src/test/resources/org/zaproxy/zap/spider/parser/html/ImgElementsSpiderHtmlParser.html:26]: No problem, been there 😃

File: zap/src/test/resources/org/zaproxy/zap/spider/parser/html/AppletElementsSpiderHtmlParser.html (hunk 5 above)
Comment (Leyart) [reply zap/src/test/resources/org/zaproxy/zap/spider/parser/html/AppletElementsSpiderHtmlParser.html:29]: You are "absolutely" right :)

File: zap/src/test/resources/org/zaproxy/zap/spider/parser/html/AppletElementsSpiderHtmlParser.html [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+<base href="http://example.com/sample/">
+<meta charset="UTF-8">
+<title>Input Elements - Spider HTML Parser</title>
+</head>
+<body>
+<!--  Ignored: -->
+<applet></applet>
+
+<!--  src -->
+<applet type="deprecated" src="/cool/applet/file.class"  alt="badpractice" code="somecode" height="320" width="480" >
```
Comment (thc202) [inline zap/src/test/resources/org/zaproxy/zap/spider/parser/html/AppletElementsSpiderHtmlParser.html:13]: Irrelevant attributes (and elements) should be removed.

File: zap/src/test/resources/org/zaproxy/zap/spider/parser/html/InputElementsSpiderHtmlParser.html (hunk 2 above)
Comment (kingthorin) [reply zap/src/test/resources/org/zaproxy/zap/spider/parser/html/InputElementsSpiderHtmlParser.html:12]: Probably a miss on our part in other files. Not a blocker, but it looks out of place to me.

File: zap/src/test/resources/org/zaproxy/zap/spider/parser/html/AppletElementsSpiderHtmlParser.html (hunk 6 above)
Comment (Leyart) [reply zap/src/test/resources/org/zaproxy/zap/spider/parser/html/AppletElementsSpiderHtmlParser.html:13]: I was adding them because my linter was complaining about missing required attributes for the tag. 
I will remove them wherever I added them for simplicity

Comment (kingthorin) [reply zap/src/test/resources/org/zaproxy/zap/spider/parser/html/AppletElementsSpiderHtmlParser.html:13]: It probably wont' matter after you tackle the other comment but there's a dbl space leading the alt attr.

If we were actually building web content that'd be fine (to include what the linter wanted) but for test purposes it's best to just focus on what's needed for the test.

**Comment by Leyart** [comment]
Thanks for the thorough review for an apparently simple PR :) Looking forward to contribute more

**Comment by kingthorin** [comment]
It's all a learning curve, no worries. Always nice to see new names.

File: zap/src/test/resources/org/zaproxy/zap/spider/parser/html/AppletElementsSpiderHtmlParser.html (hunk 6 above)
Comment (kingthorin) [reply zap/src/test/resources/org/zaproxy/zap/spider/parser/html/AppletElementsSpiderHtmlParser.html:13]: Align, height, width, alt, etc can all be dropped throughout the doc right?

Comment (Leyart) [reply zap/src/test/resources/org/zaproxy/zap/spider/parser/html/AppletElementsSpiderHtmlParser.html]: > Align, height, width, alt, etc can all be dropped throughout the doc right?

Forgot to push, the code was there :)

File: zap/src/test/resources/org/zaproxy/zap/spider/parser/html/AppletElementsSpiderHtmlParser.html [hunk 7]
```diff
@@ -0,0 +1,32 @@
+<!DOCTYPE html>
+<html>
+<head>
+<base href="http://example.com/sample/">
+<meta charset="UTF-8">
+<title>Applet Elements - Spider HTML Parser</title>
+</head>
+<body>
+<!-- Ignored: -->
+<applet></applet>
```
Comment (thc202) [inline zap/src/test/resources/org/zaproxy/zap/spider/parser/html/AppletElementsSpiderHtmlParser.html:10]: Move these to the end of the file to keep the same order in all files.

File: zap/src/test/resources/org/zaproxy/zap/spider/parser/html/AppletElementsSpiderHtmlParser.html [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+<!-- src -->
+<applet src="/cool/applet/file.class"></applet>
```
Comment (thc202) [inline zap/src/test/resources/org/zaproxy/zap/spider/parser/html/AppletElementsSpiderHtmlParser.html:13]: Include the attributes in the URLs to make it easier to check.

File: zap/src/test/resources/org/zaproxy/zap/spider/parser/html/AppletElementsSpiderHtmlParser.html (hunk 6 above)
Comment (thc202) [reply zap/src/test/resources/org/zaproxy/zap/spider/parser/html/AppletElementsSpiderHtmlParser.html]: The `type` in the `input` elements should be removed.

**Comment by thc202** [comment]
Commits should be fixed up.

**Comment by Leyart** [comment]
> 

Hope it is ok now. So for the future, everything squashed in a single commit? Can i keep multiple commits for review before squashing them?

File: zap/src/test/resources/org/zaproxy/zap/spider/parser/html/ImgElementsSpiderHtmlParser.html [hunk 9]
```diff
@@ -14,8 +14,10 @@
 <img src="">
 <img src="/img/absolute">
 <img src="ftp://img.example.com/">
-<img longdesc="/relative/longdesc">
```
Comment (psiinon) [inline zap/src/test/resources/org/zaproxy/zap/spider/parser/html/ImgElementsSpiderHtmlParser.html:17]: Why this change?

Comment (Leyart) [reply zap/src/test/resources/org/zaproxy/zap/spider/parser/html/ImgElementsSpiderHtmlParser.html:17]: To keep it consistent with the comment here: https://github.com/zaproxy/zaproxy/pull/7159#discussion_r830386703

Comment (kingthorin) [reply zap/src/test/resources/org/zaproxy/zap/spider/parser/html/ImgElementsSpiderHtmlParser.html:17]: With leading slash it’s actually absolute (from server root).

Comment (psiinon) [reply zap/src/test/resources/org/zaproxy/zap/spider/parser/html/ImgElementsSpiderHtmlParser.html:17]: Ha! I was thinking 'relative' as opposed to supplying the full URL, but I'll accept I'm wrong on this one :P

**Review by psiinon — APPROVED** [review]
Have tested against Crawl Maze and all 5 URLs found :D

**Review by kingthorin — APPROVED** [review]

**Comment by psiinon** [comment]
> Could you update the help to mention the new elements/attributes? https://github.com/zaproxy/zap-core-help/blob/5aea1349b9f4dc697e2e9cba4b5ccdea83161ac6/addOns/help/src/main/javahelp/contents/start/features/spider.html#L24

I forgot to do that in my PRs too ;) Will submit a PR for those as well..
